Exposing extended attributes in the bake stage ui

tonyrzhang opened the pull request:
Hey @duftler!

Let me know if you have any questions. Thanks!

- Added extended attributes to the bake stage for GCE and AWS.
- Based the UI on one of the ones in the manual judgement phase.

Note: Angular doesn't have support for 2-way binding for the keys within dictionary. I felt like it didn't really make sense for users to have the ability to edit a key since that's basically the equivalent of just creating a new (key, value) pair. 

[Here's](https://www.dropbox.com/s/7l7yzae11rl9wq8/Screenshot%202016-02-16%2013.18.41.png?dl=0) an image of what the end result looks like and [this](https://www.dropbox.com/s/6tfxvj5u1v0bgc8/Screenshot%202016-02-16%2013.43.28.png?dl=0) is a picture of the modal.

anotherchrisberry commented on app/scripts/modules/core/pipeline/config/stages/bake/modal/addExtendedAttribute.html:
```diff
@@ -0,0 +1,45 @@
+          <input type="text"
+                 class="form-control input-sm "
+                 ng-model="extendedAttribute.value"
+                 placeholder="enter an attribtue value"
```
s/attribtue/attribute

anotherchrisberry commented:
Should the fields be bigger in the stage itself? I don't know what kind of values are expected for these fields, so maybe not - just curious.

duftler commented on app/scripts/modules/core/pipeline/config/stages/bake/aws/bakeStage.html:
```diff
@@ -71,6 +71,41 @@ <h4 ng-if="viewState.loading" class="text-center">
                ng-model="stage.templateFileName"/>
       </stage-config-field>
+      <stage-config-field label="Extended Attributes" help-key="pipeline.config.bake.extendedAttributes"
+                          ng-if="viewState.roscoMode || stage.templateFileName">
```
I think should show the extended attributes if either `roscoMode` is enabled, or the `extendedAttributes` map has at least one entry. It doesn't have to be tied to the presence of `templateFileName` (I think this is likely copy/paste error).

duftler commented:
Would be cool to also render the extended attribute in the execution details, if specified. This is what happens with the `templateFileName`.

duftler commented:
Looks really good @tonyrzhang . Thanks for putting this together.
LGTM once the comments are addressed.

tonyrzhang commented:
@anotherchrisberry @duftler 

- Fixed the typo
- Made the fields take up the entire real estate in each row
- Fixed the ng-if conditionals

And I actually originally had made the updates to the bake status page to show all the extended attributes, but what I found was that it seemed to clutter up the UI too much. I toyed with changing it to some sort of click to see more in a modal approach. But then I realized that this was very similar to what the "Source" link currently leads leads to and does so I took it out.  What do you guys think?

duftler commented:
That argument makes sense re: showing the extended attributes in the execution ui. Maybe just an indication that there were in fact extended attributes, so the user is aware? I can live without it, just seemed like we should surface it somehow.

duftler commented:
LGTM. Chris, any concerns about merging?

duftler commented:
Needs a rebase.

duftler commented:
I've tested this with both GCE and AWS. Really happy with how this came out.

anotherchrisberry commented on app/scripts/modules/core/pipeline/config/stages/bake/aws/bakeStage.html:
```diff
@@ -71,6 +71,41 @@ <h4 ng-if="viewState.loading" class="text-center">
                ng-model="stage.templateFileName"/>
       </stage-config-field>
+      <stage-config-field label="Extended Attributes" help-key="pipeline.config.bake.extendedAttributes"
+                          ng-if="viewState.roscoMode || (stage.extendedAttributes && _.size(stage.extendedAttributes) > 0)">
```
There's a lot of logic in this `ng-if` tag. It would be good to move it into a method on the controller, e.g.
```javascript
this.showExtendedAttributes = () => $scope.viewState.roscoMode || _.size($scope.stage.extendedAttributes);
```
This would keep lodash from leaking into a scope variable and make it more testable (not that we're doing much testing of specific stages right now, but at some point we will).

anotherchrisberry replied:
Dang, @duftler merged a moment after I added my comment - I can make the change I suggested in my comment but it would be great if you could do it!

duftler replied:
Made the change in this pr: https://github.com/spinnaker/deck/pull/1982

tonyrzhang replied:
Gotcha makes sense.

duftler commented:
Sorry @anotherchrisberry (also realized that I tagged you as Chris by mistake).

duftler commented:
Closed https://github.com/spinnaker/spinnaker/issues/683.
